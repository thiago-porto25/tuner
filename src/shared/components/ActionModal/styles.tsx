import React from 'react';

import { Text } from 'react-native-paper';
import Icon from 'react-native-vector-icons/MaterialIcons';
import styled, { css } from 'styled-components/native';

export const ActionModalContainer = styled.View`
  ${({ theme }) => css`
    background-color: ${theme.colors.background};
    justify-content: space-between;
    padding: ${theme.space.sizeXxs};
    width: 100%;
    min-height: 256px;
    border-top-right-radius: 12px;
    border-top-left-radius: 12px;
  `}
`;

export const ButtonsContainer = styled.View`
  ${({ theme }) => css`
    gap: ${theme.space.sizeNano};
  `}
`;

export const TextContainer = styled.View`
  ${({ theme }) => css`
    margin-bottom: ${theme.space.sizeXxs};
    margin-top: ${theme.space.sizeXxxs};
  `}
`;

export const Description = styled(Text)``;

export const DescriptionContainer = styled.View``;

export const Title = styled(Text).attrs({
  variant: 'titleLarge',
})`
  ${({ theme }) => css`
    font-family: ${theme.family.bold};
    margin-bottom: ${theme.space.sizeNano};
  `}
`;

export const CloseButton = styled.TouchableOpacity.attrs({
  children: <Icon color="black" name="close" size={24} />,
  hitSlop: 32,
})`
  ${({ theme }) => css`
    position: absolute;
    top: ${theme.space.sizeXxxs};
    right: ${theme.space.sizeXxxs};
    z-index: 1;
  `}
`;
