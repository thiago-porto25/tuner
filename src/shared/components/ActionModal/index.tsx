import React from 'react';

import { Button, Modal, Portal } from 'react-native-paper';

import testIDs from '@/shared/constants/testIDs.constants';
import { ActionModalProps } from '@/shared/types/actionModalProps.interface';

import * as S from './styles';

export default function ActionModal({
  children,
  visible,
  contentContainerStyle = {
    bottom: 0,
    left: 0,
    right: 0,
    position: 'absolute',
  },
  dismissable,
  dismissableBackButton,
  onDismiss,
  overlayAccessibilityLabel,
  style,
  testID,
  theme,
  title,
  description,
  buttons,
  showClose,
  closeModal,
}: ActionModalProps) {
  function getDescription(desc: ActionModalProps['description']) {
    if (typeof desc === 'string') {
      return <S.Description>{desc}</S.Description>;
    }

    return desc;
  }

  function getButtons(btns: Required<ActionModalProps>['buttons']) {
    return btns.map(button => (
      <Button key={button.key} {...button}>
        {button.children}
      </Button>
    ));
  }

  return (
    <Portal>
      <Modal
        contentContainerStyle={contentContainerStyle}
        dismissable={dismissable}
        dismissableBackButton={dismissableBackButton}
        onDismiss={onDismiss}
        overlayAccessibilityLabel={overlayAccessibilityLabel}
        style={style}
        testID={testID}
        theme={theme}
        visible={visible}>
        {children || (
          <S.ActionModalContainer>
            {showClose && closeModal && (
              <S.CloseButton
                onPress={closeModal}
                testID={testIDs.ACTION_MODAL_CLOSE_BUTTON}
              />
            )}

            <S.TextContainer>
              {title && <S.Title>{title}</S.Title>}

              {description && (
                <S.DescriptionContainer>
                  {getDescription(description)}
                </S.DescriptionContainer>
              )}
            </S.TextContainer>

            {buttons && (
              <S.ButtonsContainer>{getButtons(buttons)}</S.ButtonsContainer>
            )}
          </S.ActionModalContainer>
        )}
      </Modal>
    </Portal>
  );
}
