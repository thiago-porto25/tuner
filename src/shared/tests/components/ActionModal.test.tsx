import React from 'react';

import { ButtonProps, ModalProps, Text } from 'react-native-paper';

import { fireEvent, render, screen } from '@/__tests__/test-utils';
import ActionModal from '@/shared/components/ActionModal';
import testID from '@/shared/constants/testIDs.constants';
import { ActionModalProps } from '@/shared/types/actionModalProps.interface';

describe('ActionModal', () => {
  const mockModalProps: Omit<ModalProps, 'children'> = {
    visible: true,
    contentContainerStyle: { backgroundColor: 'red' },
    dismissable: true,
    dismissableBackButton: true,
    onDismiss: jest.fn(),
    overlayAccessibilityLabel: 'Overlay Accessibility Label',
    style: { marginTop: 10 },
    testID: 'action-modal',
    theme: undefined,
  };

  const mockButtonProps: ButtonProps[] = [
    { key: 'button1', onPress: jest.fn(), children: 'Button 1' },
    { key: 'button2', onPress: jest.fn(), children: 'Button 2' },
  ];

  const mockProps: ActionModalProps = {
    ...mockModalProps,
    title: 'Modal Title',
    description: 'Modal Description',
    buttons: mockButtonProps,
    showClose: true,
    closeModal: jest.fn(),
  };

  it('renders the ActionModal component with correct props when there is no children', () => {
    render(<ActionModal {...mockProps} />);

    const modalContainer = screen.getByTestId(mockModalProps.testID!);
    const closeButton = screen.getByTestId(testID.ACTION_MODAL_CLOSE_BUTTON);
    const title = screen.getByText(mockProps.title!);
    const description = screen.getByText(String(mockProps.description));
    const button1 = screen.getByText(String(mockButtonProps[0].children));
    const button2 = screen.getByText(String(mockButtonProps[1].children));

    expect(modalContainer).toBeTruthy();
    expect(closeButton).toBeTruthy();
    expect(title).toBeTruthy();
    expect(description).toBeTruthy();
    expect(button1).toBeTruthy();
    expect(button2).toBeTruthy();
  });

  it('renders the ActionModal component with custom components if the children prop is passed', () => {
    const mockText = 'Modal Children';

    render(
      <ActionModal {...mockProps}>
        <Text>{mockText}</Text>
      </ActionModal>,
    );

    const modalContainer = screen.getByTestId(mockModalProps.testID!);
    const children = screen.getByText(mockText);

    expect(modalContainer).toBeTruthy();
    expect(children).toBeTruthy();
  });

  it('calls the onClose method when close button is pressed', () => {
    render(<ActionModal {...mockProps} />);
    const closeButton = screen.getByTestId(testID.ACTION_MODAL_CLOSE_BUTTON);

    fireEvent.press(closeButton);

    expect(mockProps.closeModal).toHaveBeenCalledTimes(1);
  });

  it('renders the ActionModal component with custom component as description if the description prop is a react node or an array of react nodes', () => {
    const mockDescriptionText = 'Modal Description';

    const mockDescription = <Text>{mockDescriptionText}</Text>;

    render(<ActionModal {...mockProps} description={mockDescription} />);

    const description = screen.getByText(mockDescriptionText);

    expect(description).toBeTruthy();
  });

  it('calls the onPress method of the buttons when they are pressed', () => {
    render(<ActionModal {...mockProps} />);
    const button1 = screen.getByText(String(mockButtonProps[0].children));
    const button2 = screen.getByText(String(mockButtonProps[1].children));

    fireEvent.press(button1);
    fireEvent.press(button2);

    expect(mockButtonProps[0].onPress).toHaveBeenCalledTimes(1);
    expect(mockButtonProps[1].onPress).toHaveBeenCalledTimes(1);
  });

  it('renders without a close button if showClose prop is false', () => {
    render(<ActionModal {...mockProps} showClose={false} />);

    const closeButton = screen.queryByTestId(testID.ACTION_MODAL_CLOSE_BUTTON);
    expect(closeButton).toBeNull();
  });

  it('renders without a close button if closeModal prop is not defined', () => {
    render(<ActionModal {...mockProps} closeModal={undefined} />);

    const closeButton = screen.queryByTestId(testID.ACTION_MODAL_CLOSE_BUTTON);
    expect(closeButton).toBeNull();
  });

  it('matches the snapshot', () => {
    render(<ActionModal {...mockProps} />);

    expect(screen.toJSON()).toMatchSnapshot();
  });
});
