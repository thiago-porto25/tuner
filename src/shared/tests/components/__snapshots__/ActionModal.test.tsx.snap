// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionModal matches the snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  />
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      accessibilityLiveRegion="polite"
      accessibilityViewIsModal={true}
      collapsable={false}
      onAccessibilityEscape={[Function]}
      pointerEvents="auto"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="action-modal"
    >
      <View
        accessibilityLabel="Overlay Accessibility Label"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        importantForAccessibility="no"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "rgba(50, 47, 55, 0.4)",
            "flex": 1,
            "opacity": 1,
          }
        }
        testID="action-modal-backdrop"
      />
      <View
        pointerEvents="box-none"
        style={
          [
            {
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "marginBottom": 0,
              "marginTop": 0,
            },
            {
              "marginTop": 10,
            },
          ]
        }
        testID="action-modal-wrapper"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "red",
              "opacity": 1,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 1,
                "width": 0,
              },
              "shadowOpacity": 0.15,
              "shadowRadius": 3,
            }
          }
          testID="action-modal-surface-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "red",
                "flex": undefined,
                "justifyContent": "center",
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.3,
                "shadowRadius": 1,
              }
            }
            testID="action-modal-surface"
          >
            <View
              style={
                {
                  "backgroundColor": "rgba(255, 251, 254, 1)",
                  "borderTopLeftRadius": 12,
                  "borderTopRightRadius": 12,
                  "justifyContent": "space-between",
                  "minHeight": 256,
                  "paddingBottom": 24,
                  "paddingLeft": 24,
                  "paddingRight": 24,
                  "paddingTop": 24,
                  "width": "100%",
                }
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                hitSlop={32}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                    "position": "absolute",
                    "right": 16,
                    "top": 16,
                    "zIndex": 1,
                  }
                }
                testID="ActionModal-CloseButton"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "black",
                        "fontSize": 24,
                      },
                      undefined,
                      {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                >
                  Óóç
                </Text>
              </View>
              <View
                style={
                  {
                    "marginBottom": 24,
                    "marginTop": 16,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "fontFamily": "Inter-Regular",
                          "fontSize": 22,
                          "fontWeight": "400",
                          "letterSpacing": 0,
                          "lineHeight": 28,
                        },
                        {
                          "fontFamily": "Inter-Bold",
                          "marginBottom": 8,
                        },
                      ],
                    ]
                  }
                >
                  Modal Title
                </Text>
                <View
                  style={{}}
                >
                  <Text
                    style={
                      [
                        {
                          "textAlign": "left",
                        },
                        {
                          "color": "rgba(28, 27, 31, 1)",
                          "fontFamily": "Inter-Regular",
                          "fontWeight": "400",
                          "letterSpacing": 0,
                        },
                        {
                          "writingDirection": "ltr",
                        },
                        {},
                      ]
                    }
                  >
                    Modal Description
                  </Text>
                </View>
              </View>
              <View
                style={
                  {
                    "gap": 8,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "backgroundColor": "transparent",
                      "borderRadius": 20,
                      "shadowColor": "#000",
                      "shadowOffset": {
                        "height": 0,
                        "width": 0,
                      },
                      "shadowOpacity": 0,
                      "shadowRadius": 0,
                    }
                  }
                  testID="button-container-outer-layer"
                >
                  <View
                    collapsable={false}
                    style={
                      {
                        "backgroundColor": "transparent",
                        "borderColor": "transparent",
                        "borderRadius": 20,
                        "borderStyle": "solid",
                        "borderWidth": 0,
                        "flex": undefined,
                        "minWidth": 64,
                        "shadowColor": "#000",
                        "shadowOffset": {
                          "height": 0,
                          "width": 0,
                        },
                        "shadowOpacity": 0,
                        "shadowRadius": 0,
                      }
                    }
                    testID="button-container"
                  >
                    <View
                      accessibilityRole="button"
                      accessibilityState={
                        {
                          "busy": undefined,
                          "checked": undefined,
                          "disabled": false,
                          "expanded": undefined,
                          "selected": undefined,
                        }
                      }
                      accessibilityValue={
                        {
                          "max": undefined,
                          "min": undefined,
                          "now": undefined,
                          "text": undefined,
                        }
                      }
                      accessible={true}
                      collapsable={false}
                      focusable={true}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        [
                          {
                            "overflow": "hidden",
                          },
                          {
                            "borderRadius": 20,
                          },
                        ]
                      }
                      testID="button"
                    >
                      <View
                        style={
                          [
                            {
                              "alignItems": "center",
                              "flexDirection": "row",
                              "justifyContent": "center",
                            },
                            undefined,
                          ]
                        }
                      >
                        <Text
                          numberOfLines={1}
                          selectable={false}
                          style={
                            [
                              {
                                "textAlign": "left",
                              },
                              {
                                "color": "rgba(28, 27, 31, 1)",
                                "writingDirection": "ltr",
                              },
                              [
                                {
                                  "fontFamily": "Inter-Medium",
                                  "fontSize": 14,
                                  "fontWeight": "400",
                                  "letterSpacing": 0.1,
                                  "lineHeight": 20,
                                },
                                [
                                  {
                                    "marginHorizontal": 16,
                                    "marginVertical": 9,
                                    "textAlign": "center",
                                  },
                                  false,
                                  {
                                    "marginHorizontal": 12,
                                  },
                                  undefined,
                                  false,
                                  {
                                    "color": "#f48225",
                                    "fontFamily": "Inter-Medium",
                                    "fontSize": 14,
                                    "fontWeight": "400",
                                    "letterSpacing": 0.1,
                                    "lineHeight": 20,
                                  },
                                  undefined,
                                ],
                              ],
                            ]
                          }
                          testID="button-text"
                        >
                          Button 1
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
                <View
                  collapsable={false}
                  style={
                    {
                      "backgroundColor": "transparent",
                      "borderRadius": 20,
                      "shadowColor": "#000",
                      "shadowOffset": {
                        "height": 0,
                        "width": 0,
                      },
                      "shadowOpacity": 0,
                      "shadowRadius": 0,
                    }
                  }
                  testID="button-container-outer-layer"
                >
                  <View
                    collapsable={false}
                    style={
                      {
                        "backgroundColor": "transparent",
                        "borderColor": "transparent",
                        "borderRadius": 20,
                        "borderStyle": "solid",
                        "borderWidth": 0,
                        "flex": undefined,
                        "minWidth": 64,
                        "shadowColor": "#000",
                        "shadowOffset": {
                          "height": 0,
                          "width": 0,
                        },
                        "shadowOpacity": 0,
                        "shadowRadius": 0,
                      }
                    }
                    testID="button-container"
                  >
                    <View
                      accessibilityRole="button"
                      accessibilityState={
                        {
                          "busy": undefined,
                          "checked": undefined,
                          "disabled": false,
                          "expanded": undefined,
                          "selected": undefined,
                        }
                      }
                      accessibilityValue={
                        {
                          "max": undefined,
                          "min": undefined,
                          "now": undefined,
                          "text": undefined,
                        }
                      }
                      accessible={true}
                      collapsable={false}
                      focusable={true}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        [
                          {
                            "overflow": "hidden",
                          },
                          {
                            "borderRadius": 20,
                          },
                        ]
                      }
                      testID="button"
                    >
                      <View
                        style={
                          [
                            {
                              "alignItems": "center",
                              "flexDirection": "row",
                              "justifyContent": "center",
                            },
                            undefined,
                          ]
                        }
                      >
                        <Text
                          numberOfLines={1}
                          selectable={false}
                          style={
                            [
                              {
                                "textAlign": "left",
                              },
                              {
                                "color": "rgba(28, 27, 31, 1)",
                                "writingDirection": "ltr",
                              },
                              [
                                {
                                  "fontFamily": "Inter-Medium",
                                  "fontSize": 14,
                                  "fontWeight": "400",
                                  "letterSpacing": 0.1,
                                  "lineHeight": 20,
                                },
                                [
                                  {
                                    "marginHorizontal": 16,
                                    "marginVertical": 9,
                                    "textAlign": "center",
                                  },
                                  false,
                                  {
                                    "marginHorizontal": 12,
                                  },
                                  undefined,
                                  false,
                                  {
                                    "color": "#f48225",
                                    "fontFamily": "Inter-Medium",
                                    "fontSize": 14,
                                    "fontWeight": "400",
                                    "letterSpacing": 0.1,
                                    "lineHeight": 20,
                                  },
                                  undefined,
                                ],
                              ],
                            ]
                          }
                          testID="button-text"
                        >
                          Button 2
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
